File: examples/showcase.pl
#!/usr/bin/perl
# Hatsune Miku Theme - Perl Showcase
# All-Miku Synthesis: Every voice, one stage.

use strict;
use warnings;
use v5.36;  # Enable modern features
use feature qw(say signatures);

# Pragmas
use utf8;
use autodie;

# Constants
use constant {
    CANONICAL_COLOR => '#39C5BB',
    DEFAULT_BPM     => 39,
    MAX_ENERGY      => 100,
};

# Package/Module definition
package MikuShowcase;

use Carp qw(croak confess);
use Scalar::Util qw(blessed reftype);
use List::Util qw(sum max min reduce);

# Class using Moo (lightweight OO)
package VoiceBank {
    use Moo;
    use Types::Standard qw(Str Int ArrayRef Bool);

    has name => (
        is       => 'ro',
        isa      => Str,
        required => 1,
    );

    has version => (
        is      => 'ro',
        isa     => Str,
        default => 'V2',
    );

    has frequency_range => (
        is      => 'ro',
        isa     => ArrayRef[Int],
        default => sub { [80, 1100] },
    );

    has active => (
        is      => 'rw',
        isa     => Bool,
        default => 1,
    );

    sub get_info {
        my ($self) = @_;
        return sprintf "%s (%s)", $self->name, $self->version;
    }
}

# Main class
package DigitalDiva {
    use Moo;
    use Types::Standard qw(Str Int InstanceOf);
    use namespace::clean;

    # Class attribute
    our $MODEL_ID = 'CV01';

    has name => (
        is      => 'ro',
        isa     => Str,
        default => 'Hatsune Miku',
    );

    has voice_bank => (
        is      => 'ro',
        isa     => InstanceOf['VoiceBank'],
        lazy    => 1,
        builder => '_build_voice_bank',
    );

    has _energy => (
        is       => 'rw',
        isa      => Int,
        default  => MAX_ENERGY,
        init_arg => 'energy',
    );

    has current_song => (
        is  => 'rw',
        isa => Str,
    );

    sub _build_voice_bank {
        my ($self) = @_;
        return VoiceBank->new(name => $self->name);
    }

    # Getter/Setter
    sub energy {
        my ($self, $value) = @_;
        if (defined $value) {
            $value = 0 if $value < 0;
            $value = MAX_ENERGY if $value > MAX_ENERGY;
            $self->_energy($value);
        }
        return $self->_energy;
    }

    sub sing {
        my ($self, $song) = @_;

        croak "Low energy: Please recharge with leeks"
            if $self->energy < 10;

        $self->energy($self->energy - 10);
        $self->current_song($song);

        return "[MIKU] Now singing: $song";
    }

    sub sing_at {
        my ($self, $song, $bpm) = @_;
        $bpm //= 120;  # Default value
        my $result = $self->sing($song);
        return "$result @ ${bpm}BPM";
    }
}

# Return to main package
package main;

# Subroutine with signature (5.36+)
sub format_song($title, $bpm = 120) {
    return "[MIKU] $title @ ${bpm}BPM";
}

# Traditional subroutine
sub format_song_legacy {
    my ($title, $bpm) = @_;
    $bpm //= 120;
    return "[MIKU] $title @ ${bpm}BPM";
}

# Anonymous subroutine
my $double = sub { $_[0] * 2 };
my $add = sub ($a, $b) { $a + $b };

# Higher-order function
sub with_logging {
    my ($fn) = @_;
    return sub {
        say "Calling function...";
        my $result = $fn->(@_);
        say "Result: $result";
        return $result;
    };
}

# String operations
sub string_showcase {
    my $name = "Hatsune Miku";

    # Interpolation
    my $greeting = "Hello, $name!";

    # Concatenation
    my $full = $name . " V2";

    # Here-doc
    my $lyrics = <<'END_LYRICS';
World is Mine
The number one princess in the world
END_LYRICS

    # Interpolated here-doc
    my $message = <<"END_MESSAGE";
Now playing: $name
Status: Active
END_MESSAGE

    # Quote operators
    my @words = qw(Melt World Love);  # No quotes needed
    my $pattern = qr/Miku[-_]?(\w+)?/i;

    # String functions
    my $upper = uc($name);
    my $lower = lc($name);
    my $len = length($name);
    my $sub = substr($name, 0, 7);

    return $greeting;
}

# Array operations
sub array_showcase {
    my @songs = ('Melt', 'World is Mine', 'Love is War');

    # Push/pop
    push @songs, 'Rolling Girl';
    my $last = pop @songs;

    # Shift/unshift
    unshift @songs, 'First Song';
    my $first = shift @songs;

    # Slice
    my @first_two = @songs[0, 1];
    my @range = @songs[0..2];

    # Map
    my @uppercased = map { uc } @songs;
    my @lengths = map { length } @songs;

    # Grep (filter)
    my @long_songs = grep { length > 5 } @songs;

    # Sort
    my @sorted = sort @songs;
    my @by_length = sort { length($a) <=> length($b) } @songs;

    # Join
    my $joined = join ', ', @songs;

    # Reduce
    my $total_length = reduce { $a + length($b) } 0, @songs;

    return \@songs;
}

# Hash operations
sub hash_showcase {
    my %config = (
        name    => 'Miku',
        version => 'V2',
        settings => {
            bpm      => 120,
            autoTune => 1,
        },
    );

    # Access
    my $name = $config{name};
    my $bpm = $config{settings}{bpm};

    # Exists/defined
    if (exists $config{name} && defined $config{name}) {
        say "Name: $name";
    }

    # Keys/values
    my @keys = keys %config;
    my @values = values %config;

    # Each (iterate)
    while (my ($key, $value) = each %config) {
        say "$key => $value" unless ref $value;
    }

    # Slice
    my @selected = @config{qw(name version)};

    # Delete
    delete $config{temp} if exists $config{temp};

    return \%config;
}

# Regular expressions
sub regex_showcase {
    my $text = "Hatsune Miku V2 - World is Mine";

    # Match
    if ($text =~ /Miku/) {
        say "Found Miku!";
    }

    # Capture groups
    if ($text =~ /(\w+)\s+(\w+)\s+(V\d+)/) {
        my ($first, $last, $version) = ($1, $2, $3);
        say "Name: $first $last, Version: $version";
    }

    # Named captures
    if ($text =~ /(?<name>Miku)\s+(?<ver>V\d+)/) {
        say "Name: $+{name}, Version: $+{ver}";
    }

    # Substitution
    my $modified = $text;
    $modified =~ s/Miku/MIKU/g;

    # Transliteration
    my $upper = $text;
    $upper =~ tr/a-z/A-Z/;

    # Split
    my @words = split /\s+/, $text;

    return \@words;
}

# File I/O
sub file_showcase {
    my $filename = 'songs.txt';

    # Write
    open my $fh, '>', $filename or die "Cannot open: $!";
    print $fh "Melt\nWorld is Mine\n";
    close $fh;

    # Read all
    open $fh, '<', $filename or die "Cannot open: $!";
    my @lines = <$fh>;
    close $fh;

    # Read line by line
    open $fh, '<', $filename or die "Cannot open: $!";
    while (my $line = <$fh>) {
        chomp $line;
        say "Song: $line";
    }
    close $fh;

    # Slurp mode
    local $/;  # Enable slurp mode
    open $fh, '<', $filename;
    my $content = <$fh>;
    close $fh;

    unlink $filename;  # Cleanup
}

# Error handling
sub safe_sing {
    my ($diva, $song) = @_;

    my $result = eval {
        $diva->sing($song);
    };

    if ($@) {
        warn "Error: $@";
        return undef;
    }

    return $result;
}

# Try::Tiny style (if available)
# use Try::Tiny;
# try {
#     $diva->sing($song);
# } catch {
#     warn "Error: $_";
# } finally {
#     say "Attempt complete";
# };

# Main
sub main {
    my $miku = DigitalDiva->new(
        name   => 'Hatsune Miku',
        energy => MAX_ENERGY,
    );

    my @songs = qw(Melt World_is_Mine Love_is_War);

    say "=== Performance Start ===";

    for my $song (@songs) {
        my $result = safe_sing($miku, $song);
        if (defined $result) {
            say $result;
        } else {
            last;  # Exit loop on error
        }
    }

    say sprintf "Final energy: %d", $miku->energy;
    say "=== Performance Complete ===";
}

main() unless caller;  # Run if executed directly

1;  # Module return value
__END__

=head1 NAME

MikuShowcase - Hatsune Miku Theme Perl Showcase

=head1 SYNOPSIS

    use MikuShowcase;
    my $miku = DigitalDiva->new(name => 'Miku');
    say $miku->sing('World is Mine');

=head1 DESCRIPTION

This module demonstrates Perl syntax highlighting features.

=head1 AUTHOR

Miku Fan

=cut

File: examples/showcase.lua
--[[
Hatsune Miku Theme - Lua Showcase
All-Miku Synthesis: Every voice, one stage.
]]

-- Module definition
local MikuShowcase = {}

-- Constants
local CANONICAL_COLOR = "#39C5BB"
local DEFAULT_BPM = 39
local MAX_ENERGY = 100

-- Version enum (as table)
local MikuVersion = {
    V2_CLASSIC = "V2",
    V3 = "V3",
    V4X = "V4X",
    NT = "NT",
    SEKAI = "SEKAI"
}

-- Simple class pattern using metatables
local VoiceBank = {}
VoiceBank.__index = VoiceBank

function VoiceBank.new(name, version, frequencyRange)
    local self = setmetatable({}, VoiceBank)
    self.name = name or "Unknown"
    self.version = version or MikuVersion.V2_CLASSIC
    self.frequencyRange = frequencyRange or {80, 1100}
    self.active = true
    return self
end

function VoiceBank:getInfo()
    return string.format("%s (%s)", self.name, self.version)
end

-- Main class
local DigitalDiva = {}
DigitalDiva.__index = DigitalDiva

-- Class constant
DigitalDiva.MODEL_ID = "CV01"

-- Constructor
function DigitalDiva.new(name, version)
    local self = setmetatable({}, DigitalDiva)
    self.name = name or "Hatsune Miku"
    self.voiceBank = VoiceBank.new(self.name, version)
    self._energy = MAX_ENERGY  -- Private by convention
    self.currentSong = nil
    return self
end

-- Getter
function DigitalDiva:getEnergy()
    return self._energy
end

-- Setter
function DigitalDiva:setEnergy(value)
    self._energy = math.max(0, math.min(MAX_ENERGY, value))
end

-- Method
function DigitalDiva:sing(song)
    if self._energy < 10 then
        error("Low energy: Please recharge with leeks")
    end
    self._energy = self._energy - 10
    self.currentSong = song
    return string.format("[MIKU] Now singing: %s", song)
end

-- Method with optional parameters
function DigitalDiva:singAt(song, bpm)
    bpm = bpm or 120
    local result = self:sing(song)
    return string.format("%s @ %dBPM", result, bpm)
end

-- Inheritance pattern
local AppendVoice = setmetatable({}, {__index = DigitalDiva})
AppendVoice.__index = AppendVoice

function AppendVoice.new(name, voiceType)
    local self = setmetatable(DigitalDiva.new(name, MikuVersion.V4X), AppendVoice)
    self.voiceType = voiceType or "sweet"
    return self
end

function AppendVoice:sing(song)
    local result = DigitalDiva.sing(self, song)
    return string.format("%s [%s]", result, self.voiceType)
end

-- Utility functions
local function formatSong(title)
    return "[MIKU] " .. title
end

-- Variadic function
local function printAll(...)
    local args = {...}
    for i, v in ipairs(args) do
        print(i, v)
    end
    return #args
end

-- Higher-order function
local function withLogging(fn)
    return function(...)
        print("Calling function...")
        local result = fn(...)
        print("Result:", result)
        return result
    end
end

-- Table operations
local function tableShowcase()
    -- Array-style table
    local songs = {"Melt", "World is Mine", "Love is War"}

    -- Insert
    table.insert(songs, "Rolling Girl")

    -- Remove
    local removed = table.remove(songs)

    -- Sort
    table.sort(songs)

    -- Concatenate
    local joined = table.concat(songs, ", ")

    -- Dictionary-style table
    local config = {
        name = "Miku",
        version = "V2",
        settings = {
            bpm = 120,
            autoTune = true
        }
    }

    -- Access
    local bpm = config.settings.bpm
    local version = config["version"]

    -- Iterate array
    for i, song in ipairs(songs) do
        print(i, song)
    end

    -- Iterate dictionary
    for key, value in pairs(config) do
        print(key, value)
    end

    return songs
end

-- String operations
local function stringShowcase()
    local title = "World is Mine"

    -- Length
    local len = #title

    -- Concatenation
    local formatted = "[MIKU] " .. title

    -- Pattern matching (Lua patterns, not regex)
    local pattern = "(%w+)%s+(%w+)"
    local word1, word2 = string.match(title, pattern)

    -- Substitution
    local replaced = string.gsub(title, "World", "Stage")

    -- Find
    local start, finish = string.find(title, "Mine")

    -- Format
    local output = string.format("Playing: %s (length: %d)", title, len)

    -- Byte/char operations
    local firstByte = string.byte(title, 1)
    local char = string.char(65)  -- 'A'

    return output
end

-- Coroutines
local function coroutineShowcase()
    local performer = coroutine.create(function(songs)
        for _, song in ipairs(songs) do
            print("Singing:", song)
            coroutine.yield(song)
        end
        return "Concert complete"
    end)

    local songs = {"A", "B", "C"}

    -- Resume coroutine
    local status, result = coroutine.resume(performer, songs)
    while coroutine.status(performer) ~= "dead" do
        status, result = coroutine.resume(performer)
        if status then
            print("Yielded:", result)
        end
    end
end

-- Producer-consumer with coroutines
local function producerConsumer()
    local function producer()
        return coroutine.create(function()
            for i = 1, 5 do
                coroutine.yield("Song " .. i)
            end
        end)
    end

    local function consumer(prod)
        while true do
            local status, value = coroutine.resume(prod)
            if not status or value == nil then break end
            print("Consumed:", value)
        end
    end

    consumer(producer())
end

-- Error handling
local function safeCall(fn, ...)
    local status, result = pcall(fn, ...)
    if status then
        return result
    else
        print("Error:", result)
        return nil
    end
end

-- xpcall with traceback
local function safeCallWithTrace(fn, ...)
    local function errorHandler(err)
        return debug.traceback(err, 2)
    end

    local args = {...}
    local status, result = xpcall(function()
        return fn(table.unpack(args))
    end, errorHandler)

    return status, result
end

-- Weak tables (for caching)
local cache = setmetatable({}, {__mode = "v"})

-- Metamethods showcase
local Vector = {}
Vector.__index = Vector

function Vector.new(x, y)
    return setmetatable({x = x, y = y}, Vector)
end

function Vector.__add(a, b)
    return Vector.new(a.x + b.x, a.y + b.y)
end

function Vector.__tostring(v)
    return string.format("(%d, %d)", v.x, v.y)
end

function Vector.__eq(a, b)
    return a.x == b.x and a.y == b.y
end

-- Module export
MikuShowcase.VoiceBank = VoiceBank
MikuShowcase.DigitalDiva = DigitalDiva
MikuShowcase.AppendVoice = AppendVoice
MikuShowcase.MikuVersion = MikuVersion
MikuShowcase.formatSong = formatSong

-- Main execution
local function main()
    local miku = DigitalDiva.new("Hatsune Miku", MikuVersion.V2_CLASSIC)
    local songs = {"Melt", "World is Mine", "Love is War"}

    print("=== Performance Start ===")

    for _, song in ipairs(songs) do
        local status, result = pcall(function()
            return miku:sing(song)
        end)

        if status then
            print(result)
        else
            print("Error:", result)
            break
        end
    end

    print(string.format("Final energy: %d", miku:getEnergy()))
    print("=== Performance Complete ===")
end

main()

return MikuShowcase
